libs/vkd3d/device.c

```diff
@@ -379,6 +379,184 @@ static bool vkd3d_disable_nvx_extensions(struct d3d12_device *device, const char
     return disabled;
 }
 
+static BOOL vkd3d_wait_vr_key(HKEY vr_key)
+{
+    DWORD type, value, wait_status, size;
+    LSTATUS status;
+    HANDLE event;
+    DWORD max_retry = 10; /* Each value query and each timeout counts as a try. */
+
+    size = sizeof(value);
+    if ((status = RegQueryValueExA(vr_key, "state", NULL, &type, (BYTE *)&value, &size)))
+    {
+        ERR("OpenVR: could not query value, status %d", status);
+        return false;
+    }
+    if (type != REG_DWORD)
+    {
+        ERR("OpenVR: unexpected value type %d", type);
+        return false;
+    }
+
+    if (value)
+        return value == 1;
+
+    event = CreateEventA(NULL, FALSE, FALSE, NULL);
+    while (max_retry)
+    {
+        if (RegNotifyChangeKeyValue(vr_key, FALSE, REG_NOTIFY_CHANGE_LAST_SET, event, TRUE))
+        {
+            ERR("Error registering registry change notification");
+            break;
+        }
+        size = sizeof(value);
+        if ((status = RegQueryValueExA(vr_key, "state", NULL, &type, (BYTE *)&value, &size)))
+        {
+            ERR("OpenVR: could not query value, status %d", status);
+            break;
+        }
+        if (value)
+            break;
+        max_retry--;
+
+        while ((wait_status = WaitForSingleObject(event, 1000)) == WAIT_TIMEOUT && max_retry)
+        {
+            WARN("VR state wait timeout (retries left %u)", max_retry);
+            max_retry--;
```

Comment:
continue after this?

Reply:
Yes.

Reply:
Why do we need that?

Reply:
I am afraid 3 seconds timeout might be a bit risky? Is there any reason not to take what dxvk has? I think that was working for some years now without any issues.

Reply:
Yes, it may take some time on first run at least (if it is installed and configured of course. The initialization in steam.exe helper is done asynchronously not to delay starting up everything which everything has a lot of things to do before first d3d device is created.

For the years of that code being in dxvk I am not aware of a single instance of hang due to that. There is an error logging during wait so it would be immediately clear if it is hanging waiting for VR [registry] initialization.

Reply:
In any case, I doubt 3 or 5 sec timeout solves anything here. If in theory we will be hitting that timeout on each [other] run and continuing after wait this is not a solution anyway, such delays on each device creation are unacceptable anyway.